### PR TITLE
Fix gufunc with multiple outputs

### LINKED ISSUE #9049: cpu `@guvectorize` doesn't support multiple outputs
## Reporting a bug

- [x] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/main/CHANGE_LOG).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

```python
from numba import guvectorize
import numpy as np

@guvectorize('(n),(n) -> (n),(n)')
def divmod(x, y, div, mod):
    for i in range(x.shape[0]):
        div[i], mod[i] = np.divmod(x[i], y[i])

x = np.arange(5, dtype='i4') + 5
y = np.full_like(x, fill_value=2)
div = np.zeros_like(x)
mod = np.zeros_like(x)
divmod(x, y, div, mod)
```

```bash
$ python reproducer.py
Traceback (most recent call last):
  File "/Users/guilhermeleobas/git/numba/reproducer.py", line 25, in <module>
    divmod(x, y, div, mod)
  File "/Users/guilhermeleobas/git/numba/numba/np/ufunc/gufunc.py", line 357, in __call__
    raise TypeError(msg)
TypeError: Too few arguments for function 'divmod'. Note that the pattern `out = gufunc(Arg1, Arg2, ..., ArgN)` is not allowed. Use `gufunc(Arg1, Arg2, ..., ArgN, out) instead.
```

This is supported in NumPy, so, it is a bug in Numba side:

```python
$ python
Python 3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:41:52) [Clang 15.0.7 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.linalg._umath_linalg.svd_n_s.signature
'(m,n)->(m,n),(n),(n,n)'
```

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Fixes #9049

### CODE REVIEW

**Comment by zaa730Sight** [comment]
Any updates on this? This is a very welcome, minimal, and straightforward fix. I use Numba in conjunction with Polars (writing UDF kernels with numba), and having the proper and working multi-out return signature is very useful.

**Comment by guilhermeleobas** [comment]
Hi @zaa730Sight, thanks for the interest in this one. This is still pending a review. I'll try to land this one for the 0.59 release.

File: numba/np/ufunc/gufunc.py [hunk 1]
```diff
@@ -158,9 +158,10 @@ def _get_signature(self, *args):
                 l.append(types.Array(ewise_types[idx], ndim, 'A'))
 
         # add return type to signature
-        retty = ewise_types[-1]
-        ret_ndim = len(parsed_sig[-1][0]) or 1  # small hack to return scalar
-        l.append(types.Array(retty, ret_ndim, 'A'))
+        for idx, sig_dim in enumerate(parsed_sig[1]):
+            retty = ewise_types[idx]
+            ret_ndim = len(sig_dim) or 1  # small hack to return scalars
+            l.append(types.Array(retty, ret_ndim, 'A'))
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/np/ufunc/gufunc.py:165]: Thanks, there was a bug in the implementation and in the tests cases. The GUFuncs needs to be dynamic in order to trigger the test case.

File: numba/np/ufunc/gufunc.py [hunk 2]
```diff
@@ -140,7 +140,7 @@ def _get_ewise_dtypes(self, args):
     def _num_args_match(self, *args):
         parsed_sig = parse_signature(self.gufunc_builder.signature)
         # parsed_sig[1] has always length 1
```
Comment (kc611) [inline numba/np/ufunc/gufunc.py:142]:
```suggestion
```

Since now `parsed_sig[1]` variable length.

File: numba/tests/npyufunc/test_gufunc.py [hunk 3]
```diff
@@ -97,6 +97,90 @@ def gufunc(x, y, res):
...
+            for i in range(B.size):
+                B[i] = A[i]
+                C[i] = A[i] * 2
+
+        A = np.arange(10, dtype=np.float32) + 1
+        B = np.zeros_like(A)
+        C = np.zeros_like(A)
+        copy_and_double(A, B, C)
+        np.testing.assert_allclose(A, B)
+        np.testing.assert_allclose(A * 2, C)
```
Comment (kc611) [inline numba/tests/npyufunc/test_gufunc.py:147]: Not sure is the allocation is being tested here. Isn't the implementation same as `test_multiple_outputs_distinct_values`?

File: numba/tests/npyufunc/test_gufunc.py [hunk 4]
```diff
@@ -97,6 +97,90 @@ def gufunc(x, y, res):
...
+            for i in range(B.size):
+                B[i] = A[i]
+                C[i] = A[i] * 1.5
+
+        A = np.arange(10, dtype=np.int32) + 1
+        B = np.zeros_like(A)
+        C = np.zeros_like(A, dtype=np.float64)
+        copy_and_multiply(A, B, C)
+        np.testing.assert_allclose(A, B)
+        np.testing.assert_allclose(A * np.float64(1.5), C)
```
Comment (kc611) [inline numba/tests/npyufunc/test_gufunc.py:163]: Might be worth validating the output `dtype`s here too

File: numba/tests/npyufunc/test_gufunc.py [hunk 5]
```diff
@@ -97,6 +97,90 @@ def gufunc(x, y, res):
...
+        # Inputs only, too few
+        msg = "Too few arguments for function 'f'"
+        with self.assertRaises(TypeError) as te:
+            f(arr)
+        self.assertIn(msg, str(te.exception))
+
+        # Inputs and outputs, too many
+        with self.assertRaises(TypeError) as te:
+            f(arr, arr, arr, arr, arr)
+        self.assertIn(msg, str(te.exception))
```
Comment (kc611) [inline numba/tests/npyufunc/test_gufunc.py:181]: Why does passing too many argument raises error which indicates too few argument?

**Review by kc611 — COMMENTED** [review]
Have the following comments.

File: numba/tests/npyufunc/test_gufunc.py (hunk 4 above)
Comment (guilhermeleobas) [reply numba/tests/npyufunc/test_gufunc.py:163]: `C` dtype won't change. Thus, I think checking the `dtype` is redundant here.

File: numba/tests/npyufunc/test_gufunc.py (hunk 3 above)
Comment (guilhermeleobas) [reply numba/tests/npyufunc/test_gufunc.py:147]: You are right. Removed this one.

FYI, I copied those tests from the `cuda` target and did not adapt them to the `cpu` target.

File: numba/tests/npyufunc/test_gufunc.py (hunk 5 above)
Comment (kc611) [reply numba/tests/npyufunc/test_gufunc.py:166]: Yep that's fine. :+1:

File: numba/tests/npyufunc/test_gufunc.py (hunk 4 above)
Comment (kc611) [reply numba/tests/npyufunc/test_gufunc.py:148]: I see

**Review by kc611 — APPROVED** [review]
LGTM!

**Pull request merged**